### PR TITLE
Switch to pipeline build

### LINKED ISSUE #97: Dockerize the build on the "possum" branch and run the specs and cukes in Jenkins

### PR DESCRIPTION
**Pull request by apotterri**
Fixes #97.

To review, verify that the new pipeline build is green.

Note that I marked two cukes `@wip`. These will be addressed in a follow on story.

### CODE REVIEW

File: lib/conjur/host_factory_token.rb [hunk 1]
```diff
@@ -66,5 +66,13 @@ def cidr
     def revoke
       Conjur::API.revoke_host_factory_token @credentials, token
     end
+
+    def ==(other)
```
Comment (apotterri) [inline lib/conjur/host_factory_token.rb]: This was an attempt to fix the failing host factory token scenario above. It's a sensible change, but wasn't sufficient to fix the problem.

File: test.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+test_image=api-test
+docker build -t $test_image -f Dockerfile.test .
+
+function finish {
+	docker rm -f $pg_cid
+	docker rm -f $server_cid
+}
+trap finish EXIT
+
+possum_tag=push-image_170626_0.1.0
```
Comment (apotterri) [inline test.sh]: This will need to be updated from `push-image_170626` to `master` after https://github.com/conjurinc/possum/pull/74 gets merged.

**Comment by dustinmm80** [comment]
Linked to issue #97

File: test.sh (hunk 2 above)
Comment (dustinmm80) [inline test.sh]: Do we need to rethink this now cuz this?

https://github.com/conjurinc/possum/pull/74#discussion_r124386003

File: test.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+mkdir -p spec/reports features/reports
+
+docker run \
+	-i \
+	--rm \
+	--link $pg_cid:pg \
+	--link $server_cid:possum \
+    -v $PWD/spec/reports:/src/spec/reports \
+    -v $PWD/features/reports:/src/features/reports \
+	-e DATABASE_URL=postgres://postgres@pg/postgres \
```
Comment (dustinmm80) [inline test.sh]: This is badly in need of a docker-compose but we can tackle that in another PR.

File: test.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash -ex
```
Comment (dustinmm80) [inline test.sh]: Do we need to use `-x` here, very verbose? We tend to vomit everything into jenkins and then it's hard to really read the logs. This is a bigger issue than this PR, but I'm curious - do you think the api-ruby build is too noisy or is it fine?

https://jenkins.conjur.net/job/api-ruby-pipeline/job/tests_170626/8/consoleFull

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+  options {
+    timestamps()
+    buildDiscarder(logRotator(numToKeepStr: '30'))
+  } 
+
+  stages {
+    stage('Test') {
+      steps {
+        sh './test.sh'
```
Comment (dustinmm80) [inline Jenkinsfile]: We can split up `test.sh` and make it into another issue for another PR. That will turn this into a real pipeline with steps. Really helps with discovery and debugging, it turns out.

File: Dockerfile.test [hunk 6]
```diff
@@ -0,0 +1,12 @@
+FROM ruby:2.2.4
```
Comment (dustinmm80) [inline Dockerfile.test]: Newer Ruby! 👍 

Glad to see it. Are there any potential regressions/odd behavior that you think this base image switch might bring up?

File: lib/conjur/host_factory_token.rb (hunk 1 above)
Comment (dustinmm80) [inline lib/conjur/host_factory_token.rb]: Yikes! Another issue to fix that, sounds like.

File: test.sh (hunk 3 above)
Comment (dustinmm80) [inline test.sh]: Concretely, this would all be easier to understand and work with in a compose setup. 

For example `--link` is deprecated, see the warning here: https://docs.docker.com/engine/userguide/networking/default_network/dockerlinks/.

We're using `--link` in a bunch of other places I know, but we need to move off it one step at a time. `docker-compose` setup can be another issue.

File: test.sh (hunk 4 above)
Comment (apotterri) [inline test.sh]: I think I prefer having `-x` set. Most of the time, when I'm looking at the logs, it's because something's gone wrong. Without `-x`, it can be hard to figure out exactly where things broke. Also, with `-x`, I can cut and paste commands to rerun them when I need to.

File: test.sh (hunk 3 above)
Comment (apotterri) [inline test.sh]: Yeah, let's do that in another PR. This basically just grabs the setup code that happens in the possum dev script, but that's a lousy way to share code....

File: Dockerfile.test (hunk 6 above)
Comment (apotterri) [inline Dockerfile.test]: No, I can't think of any. There's a decent set of specs and cukes, so I'd expect a failure somewhere in them if this was going to cause a problem

File: test.sh (hunk 3 above)
Comment (apotterri) [inline test.sh]: Also, yes, the tag for the possum image needs to change once the other PR gets merged.

File: test.sh (hunk 4 above)
Comment (dustinmm80) [inline test.sh]: Good point, let's leave it as it is then.

File: Dockerfile.test (hunk 6 above)
Comment (dustinmm80) [inline Dockerfile.test]: Okay cool, everyone loves a free upgrade
